### PR TITLE
WIP for refactor for application

### PR DESCRIPTION
**Pull request by AledWEvans**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/specification/VisitSpecification.kt [hunk 1]
```diff
@@ -7,53 +7,53 @@ import jakarta.persistence.criteria.Root
...
-          root.join<OldVisit, MutableList<Prison>>(OldVisit::prison.name).get<VisitVisitor>(
+          root.join<Visit, MutableList<Prison>>(Visit::prison.name).get<VisitVisitor>(
             Prison::code.name,
           )
         predicates.add(criteriaBuilder.equal(innerJoinFromVisitToVisitors, prisonCode))
       }
 
       startDateTime?.run {
-        predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get(OldVisit::visitStart.name), startDateTime))
+        predicates.add(criteriaBuilder.greaterThanOrEqualTo(root.get(Visit::visitStart.name), startDateTime))
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/specification/VisitSpecification.kt:36]: llink to session slot? or replace with date?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  @Query(
+    "Delete FROM application " +
+      "WHERE completed = false AND " +
+      " modifyTimestamp < now()-make_interval(minute => :expiredPeriodMinutes) AND " +
+      " reference = :applicationReference ",
+    nativeQuery = true,
+  )
+  fun deleteExpiredApplications(applicationReference: String, expiredPeriodMinutes: Int): Int
+
+  @Query(
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt:37]: not needed here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    "SELECT a.completed FROM Application a WHERE a.reference = :applicationReference",
+  )
+  fun isApplicationCompleted(applicationReference: String): Boolean
+
+  @Query(
+    "SELECT a.restriction AS visitRestriction, COUNT(v) AS count  FROM application a " +
+      "JOIN session_slot ss ON ss.id = a.session_slot_id " +
+      "WHERE ss.session_template_reference = :sessionTemplateReference AND " +
+      "(ss.slotDate >= :sessionDate AND ss.slotDate < (CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL))) AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt:59]: = :sessionDate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    sessionDate: LocalDate,
+    expiredDateAndTime: LocalDateTime,
+  ): List<VisitRestrictionStats>
+
+  @Query(
+    "SELECT CASE WHEN (COUNT(a) > 0) THEN TRUE ELSE FALSE END FROM Application a " +
+      "WHERE a.completed = false AND a.reservedSlot = true AND " +
+      "(a.prisonerId = :prisonerId) AND " +
+      "(a.sessionSlot.sessionTemplateReference = :sessionTemplateReference) AND " +
+      "(a.sessionSlot.slotDate >= :startDateTime) ",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt:77]: = :slotDate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionSlotRepository.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  @Query(
+    "SELECT s FROM SessionSlot s " +
+      "WHERE s.sessionTemplateReference = :sessionTemplateReference" +
+      " AND s.slotDate = :slotDate AND s.slotTime = :slotTime AND s.slotEndTime = :slotEndTime",
+  )
+  fun findSessionSlot(
+    sessionTemplateReference: String,
+    slotDate: LocalDate,
+    slotTime: LocalTime,
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionSlotRepository.kt:22]: do we need the slot time if we got the sessiontemplatereference?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 6]
```diff
@@ -100,9 +85,9 @@ interface VisitRepository : JpaRepository<OldVisit, Long>, JpaSpecificationExecu
 
   @Query(
     "SELECT v.visit_restriction AS visitRestriction, COUNT(v) AS count  FROM visit v " +
-      "JOIN session_template st ON st.reference = v.session_template_reference " +
-      "WHERE v.session_template_reference = :sessionTemplateReference AND " +
-      "(v.visit_start >= :sessionDate AND v.visit_start < (CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL))) AND " +
+      "JOIN session_slot ss ON ss.id = v.session_slot_id " +
+      "WHERE ss.session_template_reference = :sessionTemplateReference AND " +
+      "(ss.slotDate >= :sessionDate AND ss.slotDate < (CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL))) AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:90]: ss.slotDate = sessionDate

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 7]
```diff
@@ -115,46 +100,41 @@ interface VisitRepository : JpaRepository<OldVisit, Long>, JpaSpecificationExecu
...
     "SELECT v.visit_restriction AS visitRestriction, COUNT(v) AS count  FROM visit v " +
-      "JOIN session_template st ON st.reference = v.session_template_reference " +
-      "WHERE (v.visit_start >= :sessionDate AND v.visit_start < (CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL))) AND " +
-      "v.session_template_reference = :sessionTemplateReference AND " +
-      "v.visit_restriction IN ('OPEN','CLOSED') AND " +
-      "v.visit_status = 'RESERVED' AND " +
-      "v.modify_timestamp >= :expiredDateAndTime " +
+      "JOIN session_slot ss ON ss.id = v.session_slot_id " +
+      "WHERE ss.session_template_reference = :sessionTemplateReference AND " +
+      "(ss.slotDate >= :sessionDate AND ss.slotDate < (CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL))) AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:105]: ss.slotDate = :sessionDate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 8]
```diff
@@ -233,4 +213,39 @@ interface VisitRepository : JpaRepository<OldVisit, Long>, JpaSpecificationExecu
...
+    "SELECT v FROM Visit v JOIN v.visitors vis WHERE " +
+      "(:#{#filter.prisonerId} is null OR v.prisonerId = :#{#filter.prisonerId})  AND " +
+      "(:#{#filter.prisonCode} is null OR v.prison.code = :#{#filter.prisonCode}) AND " +
+      "(:#{#filter.visitStatusList} is null OR :#{#filter.visitStatusList.size()==0} OR v.visitStatus in :#{#filter.visitStatusList}) AND " +
+      "(:#{#filter.visitorId} is null OR vis.nomisPersonId = :#{#filter.visitorId}) AND " +
+      "(:#{#filter.startDateTime} is null OR (v.sessionSlot.slotDate >= :#{#filter.startDateTime.toLocalDate()} AND v.sessionSlot.slotTime >= :#{#filter.startDateTime.toLocalTime()})) AND " +
+      "(:#{#filter.endDateTime} is null OR (v.sessionSlot.slotDate <= :#{#filter.endDateTime.toLocalDate()} AND v.sessionSlot.slotEndTime < :#{#filter.endDateTime.toLocalTime()})) " +
+      " ORDER BY v.sessionSlot.slotDate,v.sessionSlot.slotTime",
+  )
+  fun findVisitsOrderByDateAndTime(filter: VisitFilter, pageable: Pageable): Page<Visit>
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:227]: we could revisit the code and see if we need to keep the queries for start date and time and end date and time or we can replace it with a date

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt [hunk 9]
```diff
@@ -57,56 +61,149 @@ class MigrateVisitService(
...
       visitRoom = migrateVisitRequest.visitRoom
+
+      sessionSlot = sessionSlotService.getSessionSlot(
+        startTimeDate = migrateVisitRequest.startTimestamp,
+        endTimeAndDate = migrateVisitRequest.endTimestamp,
+        prison,
+      )
+    }
+
+    val applicationEntity: Application = createApplication(migrateVisitRequest, prison, sessionSlot, actionedBy)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt:108]: do we need to create an application for migrated visits?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt [hunk 10]
```diff
@@ -330,11 +330,19 @@ class SessionService(
...
   private fun sessionHasBooking(session: VisitSessionDto, prisonerId: String): Boolean {
-    return visitRepository.hasVisits(
-      prisonCode = session.prisonCode,
+    if (visitRepository.hasVisits(
+        prisonerId = prisonerId,
+        sessionTemplateReference = session.sessionTemplateReference,
+        slotDate = session.startTimestamp.toLocalDate(),
+      )
+    ) {
+      return true
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionService.kt:339]: replace with OR applicationRepository.hasReservations

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/SessionSlotDto.kt [hunk 11]
```diff
@@ -8,24 +8,21 @@ import jakarta.validation.constraints.NotNull
...
 import uk.gov.justice.digital.hmpps.visitscheduler.controller.validators.VisitorCountValidation
 import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitRestriction
-import java.time.LocalDateTime
+import java.time.LocalDate
 
-data class ReserveVisitSlotDto(
+data class SessionSlotDto(
   @Schema(description = "Prisoner Id", example = "AF34567G", required = true)
   @field:NotBlank
   val prisonerId: String,
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/SessionSlotDto.kt:16]: prisonerId is existing code, trying to reduce change on api

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/specification/VisitSpecification.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/specification/VisitSpecification.kt]: This code will be removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt:51]: Not sure what you are saying

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/SessionSlotDto.kt [hunk 12]
```diff
@@ -8,24 +8,21 @@ import jakarta.validation.constraints.NotNull
 import uk.gov.justice.digital.hmpps.visitscheduler.controller.validators.VisitorContactValidation
 import uk.gov.justice.digital.hmpps.visitscheduler.controller.validators.VisitorCountValidation
 import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitRestriction
-import java.time.LocalDateTime
+import java.time.LocalDate
 
-data class ReserveVisitSlotDto(
+data class SessionSlotDto(
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/SessionSlotDto.kt:13]: Yeha will change this to something like InitialApplicationCreateRequest

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt (hunk 9 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt:108]: Yet's keep changes to minimal...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 8 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:227]: I think get it all running revisit this when tests running.

**Pull request merged**